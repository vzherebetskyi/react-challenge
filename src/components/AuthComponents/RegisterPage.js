import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { createUserWithEmailAndPassword, updateProfile } from 'firebase/auth';

import AuthComponentWrapper from './AuthComponentWrapper';
import { auth } from '../../firebase/firebase';
import { handleChangeEmail } from '../../actions/auth';
import { handleUpdateUserProfile } from '../../actions/userProfile';
import { handleShowNotification } from '../../actions/notifications';
import {
  errorMessages,
  validEmailRegexp,
  validStrongPassword,
} from '../../utils/constants';
import { useEnterListener } from '../../customHooks/useBtnListener';
import Mail from '../../assets/images/mail.svg';
import UserImg from '../../assets/images/user.svg';
import PasswordInput from '../reusedComponents/PasswordInput';

const RegisterPage = props => {
  const dispatch = useDispatch();
  const userEmail = useSelector(state => state.auth.userEmail);
  const enterPressed = useEnterListener();

  const [credentials, setCredentials] = useState({
    username: '',
    email: userEmail,
    password: '',
    confirmpassw: '',
  });
  const [errors, setErrors] = useState({
    username: '',
    email: '',
    password: '',
    confirmpassw: '',
  });
  const [useFirebase, setUseFirebase] = useState(false);
  const [termsOfServiceAgreed, setTermsOfServiceAgreed] = useState(false);

  useEffect(() => {
    if (enterPressed && termsOfServiceAgreed) {
      handleValidate();
    }
  }, [enterPressed]);

  useEffect(() => {
    if (credentials.confirmpassw.length > 0) {
      if (
        credentials['password'] !== credentials['confirmpassw'] &&
        errors['confirmpassw'] !== errorMessages.passwordsEquality
      ) {
        const newErrors = { ...errors };
        newErrors['confirmpassw'] = errorMessages.passwordsEquality;
        setErrors(newErrors);
      } else if (credentials['password'] === credentials['confirmpassw']) {
        const newErrors = { ...errors };
        newErrors['confirmpassw'] = '';
        setErrors(newErrors);
      }
    }
  }, [credentials.password, credentials.confirmpassw]);

  const handleChangeInput = val => {
    const newCredentials = { ...credentials };
    const name = val.name;
    newCredentials[name] = val.value;
    setCredentials(newCredentials);
    if (name === 'email') {
      dispatch(handleChangeEmail(val.value));
    }
    if (errors[name] && name !== 'confirmpassw') {
      const newErrors = { ...errors, [name]: '' };
      setErrors(newErrors);
    }
  };

  const switchToLogin = () => {
    props.history.push('/');
  };

  const handleSignUp = () => {
    createUserWithEmailAndPassword(
      auth,
      credentials.email,
      credentials.password
    )
      .then(userCredential => {
        console.log('success', userCredential);
        dispatch(handleShowNotification('Signed up successfully'));

        return updateProfile(auth.currentUser, {
          displayName: credentials.username,
        });
      })
      .then(() => {
        console.log('User profile successfully updated');
        dispatch(
          handleUpdateUserProfile('username', auth.currentUser.displayName)
        );
      })
      .catch(error => {
        console.log(error.message);
        dispatch(handleShowNotification(error.message));
      });
  };

  const handleValidate = () => {
    let errorIsPresent;
    const newErrors = { ...errors };
    for (const el in credentials) {
      if (credentials[el].length === 0) {
        errorIsPresent = 1;
        newErrors[el] = errorMessages.notEmptyField;
      } else if (credentials[el].length !== 0) {
        if (el === 'username' && credentials[el].length < 4) {
          errorIsPresent = 1;
          newErrors[el] = errorMessages.usernameLength;
        }
        if (el === 'email') {
          const result = credentials[el].match(validEmailRegexp);
          if (!result) {
            errorIsPresent = 1;
            newErrors[el] = errorMessages.emailValidity;
          }
        }
        if (el === 'password' || el === 'confirmpassw') {
          const result = credentials[el].match(validStrongPassword);
          if (!result) {
            errorIsPresent = 1;
            newErrors[el] = errorMessages.passwStrengthRequirements;
          }
        }
        if (
          el === 'confirmpassw' &&
          credentials[el] !== credentials['password']
        ) {
          errorIsPresent = 1;
          newErrors[el] = errorMessages.passwordsEquality;
        }
      }
    }
    setErrors(newErrors);
    if (!errorIsPresent) {
      if (useFirebase) {
        handleSignUp();
      } else
        dispatch(
          handleShowNotification(
            'The in app registration is currently not possible'
          )
        );
    }
  };

  return (
    <AuthComponentWrapper title="SIGN UP" history={props.history}>
      <form
        onKeyPress={e => e.key === 'Enter' && e.preventDefault()}
        onSubmit={e => {
          e.preventDefault();
          handleValidate();
        }}
      >
        <div className="input-container">
          <img src={UserImg} alt="user" />
          <input
            autoComplete="off"
            placeholder="Username"
            type="text"
            name="username"
            value={credentials.username}
            onChange={e =>
              handleChangeInput({ name: e.target.name, value: e.target.value })
            }
          />
          {errors.username && (
            <p className="error-message">{errors.username}</p>
          )}
        </div>
        <div className="input-container">
          <img src={Mail} alt="mail" />
          <input
            autoComplete="off"
            placeholder="E-Mail Address"
            type="text"
            name="email"
            value={credentials.email}
            onChange={e =>
              handleChangeInput({ name: e.target.name, value: e.target.value })
            }
          />
          {errors.email && <p className="error-message">{errors.email}</p>}
        </div>
        <PasswordInput
          handleChangeInput={handleChangeInput}
          credentials={credentials}
          errors={errors}
          inputName="password"
          inputPlaceholder="Password"
        />
        <PasswordInput
          handleChangeInput={handleChangeInput}
          credentials={credentials}
          errors={errors}
          inputName="confirmpassw"
          inputPlaceholder="Confirm password"
        />
        <div className="checkbox-container" style={{ marginBottom: 0 }}>
          <input
            type="checkbox"
            id="termsOfService"
            name="termsOfService"
            checked={termsOfServiceAgreed}
            onChange={() => setTermsOfServiceAgreed(prevState => !prevState)}
          />
          <label htmlFor="termsOfService">
            I have read and agree with{' '}
            <a target="_blank" href="/terms_of_service">
              Terms of Service
            </a>
          </label>
        </div>
        <div className="checkbox-container">
          <input
            type="checkbox"
            id="useFirebase"
            name="useFirebase"
            checked={useFirebase}
            onChange={() => setUseFirebase(prevState => !prevState)}
          />
          <label htmlFor="useFirebase">
            I would like to use Firebase for registration
          </label>
        </div>
        <div className="parall-styled-btn-container">
          <div
            onClick={!termsOfServiceAgreed ? null : handleValidate}
            className={`parall-styled-btn ${
              !termsOfServiceAgreed ? 'pntr-events-none' : ''
            }`}
          >
            <button type="submit" disabled={!termsOfServiceAgreed}>
              Sign Up
            </button>
          </div>
        </div>
      </form>
      <div className="auth-container-bottom-text">
        <p onClick={switchToLogin}>
          <span>Back to Login</span>
        </p>
      </div>
    </AuthComponentWrapper>
  );
};

export default RegisterPage;
