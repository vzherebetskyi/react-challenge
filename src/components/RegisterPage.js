import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { createUserWithEmailAndPassword, updateProfile } from 'firebase/auth';

import { auth } from '../firebase/firebase';
import { handleChangeEmail } from '../actions/auth';
import { handleUpdateUserProfile } from '../actions/userProfile';
import { handleShowNotification } from '../actions/notifications';
import {
  errorMessages,
  validEmailRegexp,
  validStrongPassword,
} from '../utils/constants';
import { useEnterListener } from '../customHooks/useBtnListener';

const RegisterPage = props => {
  const dispatch = useDispatch();
  const userEmail = useSelector(state => state.auth.userEmail);
  const enterPressed = useEnterListener();

  const [credentials, setCredentials] = useState({
    username: '',
    email: userEmail,
    password: '',
    confirmpassw: '',
  });
  const [errors, setErrors] = useState({
    username: '',
    email: '',
    password: '',
    confirmpassw: '',
  });
  const [useFirebase, setUseFirebase] = useState(false);
  const [termsOfServiceAgreed, setTermsOfServiceAgreed] = useState(false);

  useEffect(() => {
    if (enterPressed && termsOfServiceAgreed) {
      handleValidate();
    }
  }, [enterPressed]);

  useEffect(() => {
    if (credentials.confirmpassw.length > 0) {
      if (
        credentials['password'] !== credentials['confirmpassw'] &&
        errors['confirmpassw'] !== errorMessages.passwordsEquality
      ) {
        const newErrors = { ...errors };
        newErrors['confirmpassw'] = errorMessages.passwordsEquality;
        setErrors(newErrors);
      } else if (credentials['password'] === credentials['confirmpassw']) {
        const newErrors = { ...errors };
        newErrors['confirmpassw'] = '';
        setErrors(newErrors);
      }
    }
  }, [credentials.password, credentials.confirmpassw]);

  const handleChangeInput = val => {
    const newCredentials = { ...credentials };
    const name = val.name;
    newCredentials[name] = val.value;
    setCredentials(newCredentials);
    if (name === 'email') {
      dispatch(handleChangeEmail(val.value));
    }
    if (errors[name] && name !== 'confirmpassw') {
      const newErrors = { ...errors, [name]: '' };
      setErrors(newErrors);
    }
  };

  const switchToLogin = () => {
    props.history.push('/');
  };

  const handleSignUp = () => {
    createUserWithEmailAndPassword(
      auth,
      credentials.email,
      credentials.password
    )
      .then(userCredential => {
        console.log('success', userCredential);
        dispatch(handleShowNotification('Signed up successfully'));

        return updateProfile(auth.currentUser, {
          displayName: credentials.username,
        });
      })
      .then(() => {
        console.log('User profile successfully updated');
        dispatch(
          handleUpdateUserProfile('username', auth.currentUser.displayName)
        );
      })
      .catch(error => {
        console.log(error.message);
        dispatch(handleShowNotification(error.message));
      });
  };

  const handleValidate = () => {
    let errorIsPresent;
    const newErrors = { ...errors };
    for (const el in credentials) {
      if (credentials[el].length === 0) {
        errorIsPresent = 1;
        newErrors[el] = errorMessages.notEmptyField;
      } else if (credentials[el].length !== 0) {
        if (el === 'username' && credentials[el].length < 4) {
          errorIsPresent = 1;
          newErrors[el] = errorMessages.usernameLength;
        }
        if (el === 'email') {
          const result = credentials[el].match(validEmailRegexp);
          if (!result) {
            errorIsPresent = 1;
            newErrors[el] = errorMessages.emailValidity;
          }
        }
        if (el === 'password' || el === 'confirmpassw') {
          const result = credentials[el].match(validStrongPassword);
          if (!result) {
            errorIsPresent = 1;
            newErrors[el] = errorMessages.passwStrengthRequirements;
          }
        }
        if (
          el === 'confirmpassw' &&
          credentials[el] !== credentials['password']
        ) {
          errorIsPresent = 1;
          newErrors[el] = errorMessages.passwordsEquality;
        }
      }
    }
    setErrors(newErrors);
    if (!errorIsPresent) {
      if (useFirebase) {
        handleSignUp();
      } else
        dispatch(
          handleShowNotification(
            'The in app registration is currently not possible'
          )
        );
    }
  };

  return (
    <form
      onKeyPress={e => e.key === 'Enter' && e.preventDefault()}
      onSubmit={e => {
        e.preventDefault();
        handleValidate();
      }}
    >
      <input
        autoComplete="off"
        placeholder="Username"
        type="text"
        name="username"
        value={credentials.username}
        onChange={e =>
          handleChangeInput({ name: e.target.name, value: e.target.value })
        }
      />
      {errors.username && <p>{errors.username}</p>}
      <input
        autoComplete="off"
        placeholder="E-Mail Address"
        type="text"
        name="email"
        value={credentials.email}
        onChange={e =>
          handleChangeInput({ name: e.target.name, value: e.target.value })
        }
      />
      {errors.email && <p>{errors.email}</p>}
      <input
        autoComplete="off"
        placeholder="Password"
        type="text"
        name="password"
        value={credentials.password}
        onChange={e =>
          handleChangeInput({ name: e.target.name, value: e.target.value })
        }
      />
      {errors.password && <p>{errors.password}</p>}
      <input
        autoComplete="off"
        placeholder="Confirm password"
        type="text"
        name="confirmpassw"
        value={credentials.confirmpassw}
        onChange={e =>
          handleChangeInput({ name: e.target.name, value: e.target.value })
        }
      />
      {errors.confirmpassw && <p>{errors.confirmpassw}</p>}
      <div>
        <input
          type="checkbox"
          id="termsOfService"
          name="termsOfService"
          checked={termsOfServiceAgreed}
          onChange={() => setTermsOfServiceAgreed(prevState => !prevState)}
        />
        <label htmlFor="termsOfService">
          I have read and agree with{' '}
          <a target="_blank" href="/terms_of_service">
            Terms of Service
          </a>
        </label>
      </div>
      <div>
        <input
          type="checkbox"
          id="useFirebase"
          name="useFirebase"
          checked={useFirebase}
          onChange={() => setUseFirebase(prevState => !prevState)}
        />
        <label htmlFor="useFirebase">
          I would like to use Firebase for registration
        </label>
      </div>
      <button type="submit" disabled={!termsOfServiceAgreed}>
        Sign Up
      </button>
      <button type="button" onClick={switchToLogin}>
        Back to Login
      </button>
    </form>
  );
};

export default RegisterPage;
