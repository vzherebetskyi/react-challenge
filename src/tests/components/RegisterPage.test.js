import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import RegisterPage from '../../components/AuthComponents/RegisterPage';

afterEach(() => {
  cleanup();
  jest.resetAllMocks();
});

jest.mock('firebase/app', () => {
  return {
    initializeApp: jest.fn(),
  };
});

jest.mock('firebase/auth', () => {
  return {
    getAuth: jest.fn(),
  };
});

const mockDispatch = jest.fn(() => jest.fn());
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

const history = { push: jest.fn() };

test('should render RegisterPage correctly', () => {
  const elem = renderer.create(<RegisterPage history={history} />).toJSON();
  expect(elem).toMatchSnapshot();
});

test('sign up button should be disabled', () => {
  render(<RegisterPage history={history} />);
  const signupBtn = screen.getByTestId('signup-btn');
  expect(signupBtn.disabled).toBe(true);
});

test('sign up button should be disabled', () => {
  render(<RegisterPage history={history} />);
  const checkboxTermsOfService = screen.getByTestId('terms-of-service');
  fireEvent.click(checkboxTermsOfService);
  const signupBtn = screen.getByTestId('signup-btn');
  expect(signupBtn.disabled).toBe(false);
});

test('sign up button click should not trigger dispatch and errors should appear (basic validation)', () => {
  const { getAllByText } = render(<RegisterPage history={history} />);
  const checkboxTermsOfService = screen.getByTestId('terms-of-service');
  fireEvent.click(checkboxTermsOfService);
  const signupBtn = screen.getByTestId('signup-btn');
  fireEvent.click(signupBtn);
  expect(mockDispatch).toHaveBeenCalledTimes(0);
  expect(getAllByText('This field should not be empty'));
});

test('errors should appear (more complex validation)', () => {
  const { container, getByText } = render(<RegisterPage history={history} />);
  const loginBtn = screen.getByTestId('signup-btn');
  const inputs = container.querySelectorAll('input');
  fireEvent.change(inputs[0], { target: { value: 'tes' } });
  fireEvent.change(inputs[1], { target: { value: 'test' } });
  fireEvent.change(inputs[2], { target: { value: 'testtest' } });
  fireEvent.change(inputs[3], { target: { value: 'testtest' } });
  const checkboxTermsOfService = screen.getByTestId('terms-of-service');
  fireEvent.click(checkboxTermsOfService);
  fireEvent.click(loginBtn);
  expect(getByText('Username should be at least 4 characters length'));
  expect(getByText('Please enter a valid email'));
});

test('passwords equality error should appear', () => {
  const { container, getByText } = render(<RegisterPage history={history} />);
  const inputs = container.querySelectorAll('input');
  fireEvent.change(inputs[2], { target: { value: 'a' } });
  fireEvent.change(inputs[3], { target: { value: 'b' } });
  expect(getByText('Password confirmation should be equal to password'));
});
