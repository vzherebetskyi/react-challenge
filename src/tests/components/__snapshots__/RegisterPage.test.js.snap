// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RegisterPage correctly 1`] = `
<div
  className="auth-container"
>
  <div
    className="auth-title-panel"
  >
    <img
      alt="main-logo"
      src={Object {}}
    />
  </div>
  <div
    className="auth-faq-button"
    onClick={[Function]}
  >
    <img
      alt="question"
      src={Object {}}
    />
  </div>
  <div
    className="form-container"
  >
    <h1
      className="base-title"
    >
      SIGN UP
    </h1>
    <form
      onKeyPress={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="input-container"
      >
        <img
          alt="username"
          src={Object {}}
        />
        <input
          autoComplete="off"
          data-testid="base-input"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
        
      </div>
      <div
        className="input-container"
      >
        <img
          alt="email"
          src={Object {}}
        />
        <input
          autoComplete="off"
          data-testid="base-input"
          name="email"
          onChange={[Function]}
          placeholder="E-Mail Address"
          type="text"
          value=""
        />
        
      </div>
      <div
        className="input-container"
      >
        <img
          alt="lock"
          src={Object {}}
        />
        <img
          alt="eye"
          className="input-container__eye"
          onClick={[Function]}
          src={Object {}}
        />
        <input
          autoComplete="off"
          data-testid="password-input"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
        
      </div>
      <div
        className="input-container"
      >
        <img
          alt="lock"
          src={Object {}}
        />
        <img
          alt="eye"
          className="input-container__eye"
          onClick={[Function]}
          src={Object {}}
        />
        <input
          autoComplete="off"
          data-testid="password-input"
          name="confirmpassw"
          onChange={[Function]}
          placeholder="Confirm password"
          type="password"
        />
        
      </div>
      <div
        className="checkbox-container"
        style={
          Object {
            "marginBottom": 0,
          }
        }
      >
        <input
          checked={false}
          id="termsOfService"
          name="termsOfService"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          data-testid="terms-of-service"
          htmlFor="termsOfService"
        >
          I have read and agree with
           
          <a
            href="/terms_of_service"
            target="_blank"
          >
            Terms of Service
          </a>
        </label>
      </div>
      <div
        className="checkbox-container"
      >
        <input
          checked={false}
          id="useFirebase"
          name="useFirebase"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="useFirebase"
        >
          I would like to use Firebase for registration
        </label>
      </div>
      <div
        className="parall-styled-btn-container"
      >
        <button
          className="parall-styled-btn pntr-events-none"
          data-testid="signup-btn"
          disabled={true}
          type="submit"
        >
          <div>
            Sign Up
          </div>
        </button>
      </div>
    </form>
    <div
      className="auth-container-bottom-text"
    >
      <p
        onClick={[Function]}
      >
        <span>
          Back to Login
        </span>
      </p>
    </div>
  </div>
</div>
`;
